logstash-core/src/main/java/org/logstash/instrument/witness/PluginsWitness.java

```diff
@@ -0,0 +1,99 @@
+package org.logstash.instrument.witness;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Witness for the set of plugins.
+ */
+public class PluginsWitness{
+
+    private final Map<String, PluginWitness> inputs;
+    private final Map<String, PluginWitness> outputs;
+    private final Map<String, PluginWitness> filters;
+    private final Forgetter forgetter;
+
+    /**
+     * Constructor.
+     */
+    public PluginsWitness() {
+
+        this.inputs = new HashMap<>();
```

Comment:
> We will be accessing these maps across threads right?

Yup, but almost all get. Only a single thread should ever update the map concurrently, but will change to CHM if that ever changes (or I am wrong). 

I will also change the lazy instantiation to use  `computeIfAbsent` to avoid the race condition.


> to have the constructor take the full immutable maps of plugins as arguments

This would require `PipelineWitness` to hold the set of `PluginWitness`. Imo that breaks some encapsulation. (`PluginsWitness` is the best place to house the set of `PluginWitness`).  Also see my comment about wanting to only construct the `PluginWitness` after we know the ID.